api/v1/verticadb_webhook.go

```diff
@@ -1765,6 +1769,85 @@ func (v *VerticaDB) checkNewSBoxOrSClusterShutdownUnset(allErrs field.ErrorList)
+// live in a sandbox where Shutdown is set to false in spec/status for all pieces
+func (v *VerticaDB) checkShutdownForScaleUpOrDown(oldObj *VerticaDB, allErrs field.ErrorList) field.ErrorList {
+	newSclusterMap := v.GenSubclusterMap()
+	newSclusterSboxMap := v.GenSubclusterSandboxMap()
+	newSandboxMap := v.GenSandboxMap()
+	statusSclusterIndexMap := v.GenStatusSClusterIndexMap()
+	oldSclusterMap := oldObj.GenSubclusterMap()
+	newSclusterIndexMap := v.GenSubclusterIndexMap()
+	for subclusterName, sandboxName := range newSclusterSboxMap {
+		oldScluster, foundSclusterInOld := oldSclusterMap[subclusterName]
+		if !foundSclusterInOld {
+			continue // we only need to deal with persistent subclusters
+		}
+		newScluster, foundSclusterInNew := newSclusterMap[subclusterName]
+		if !foundSclusterInNew {
+			continue // avoid unexpected panic when something is wrong
+		}
+		sandbox := newSandboxMap[sandboxName]
+		if oldScluster.Size != newScluster.Size { // scale up/down
+			errMsgs := v.checkSboxForShutdown(sandbox, newSclusterMap, statusSclusterIndexMap)
+			if len(errMsgs) != 0 {
+				i := newSclusterIndexMap[subclusterName]
+				p := field.NewPath("spec").Child("subclusters").Index(i).Child("size")
+				err := field.Invalid(p,
+					newScluster.Size,
+					fmt.Sprintf("cannot scale up/down subcluster %q in sandbox %q because %q",
+						subclusterName, sandboxName, strings.Join(errMsgs, ",")))
+				allErrs = append(allErrs, err)
+			}
+		}
+	}
+	return allErrs
+}
+
+// checkSClusterToBeSandboxedShutdownUnset ensures the following when a subcluster (A) is to be added to a sandbox (B)
+// 1 A's Shutdown field is set to false
+// 2 B's Shutdown field is set to false
+// 3 B does not have any subcluster in it that has Shutdown set to true (spec/status)
+func (v *VerticaDB) checkSClusterToBeSandboxedShutdownUnset(allErrs field.ErrorList) field.ErrorList {
+	newSclusterMap := v.GenSubclusterMap()
+	statusScluterSboxMap := v.GenSubclusterSandboxStatusMap()
+	newSclusterSboxMap := v.GenSubclusterSandboxMap()
+	newSandboxIndexMap := v.GenSandboxIndexMap()
+	newSandboxMap := v.GenSandboxMap()
+	statusSclusterIndexMap := v.GenStatusSClusterIndexMap()
+	sandboxWithError := map[string]bool{}
+	for subclusterName, sandboxName := range newSclusterSboxMap {
+		oldSandboxName, isInOldSandbox := statusScluterSboxMap[subclusterName]
+		_, foundSubcluster := newSclusterMap[subclusterName]
+		if !foundSubcluster {
+			continue // avoid panic. this error should have been reported by other functions
+		}
+		// the subcluster to be sandboxed is either a new subcluster to be added (not found in status)
+		// or an existing subcluster (found in status), the latter of which includes two scenarios:
+		//   1 the subcluster was in a sandbox (whose name is different from current sandbox name)
+		//   2 the subcluster was not in a sandbox previously
+		if !isInOldSandbox || isInOldSandbox && oldSandboxName != sandboxName {
```

Comment:
Use parenthesis to separate the conditions for readability.

Reply:
The case `isInOldSandbox && oldSandboxName != sandboxName` is for unsandboxing. I remember this is checked in another rule. Will we get two errors for this case from two different rules?

Reply:
Good memory! Here is the related rule from my previous PR:
You cannot unsandbox a subcluster with `shutdown` true(spec and/or status) or part of a sandbox with `shutdown` true 

There will be two error messages:
When a subcluster is moved from one sandbox to another,
-  this related rule will raise an error for **unsandboxing** scenario.
- The rule implmented in this PR will raise another error messages for **sandboxing** scenario.

Reply:
Removed this check "isInOldSandbox && oldSandboxName != sandboxName" per offline discussion to avoid duplicate error messages.